File: src/native/libs/System.Native.Browser/native/timer.c
// Licensed to the .NET Foundation under one or more agreements.
// The .NET Foundation licenses this file to you under the MIT license.

#include <emscripten.h>
#include <assert.h>

// this points to System.Threading.TimerQueue.TimerHandler C# method
static void (*timer_handler)() = NULL;

void
SystemJS_ExecuteTimerCallback (void)
{
	// callback could be null if timer was never used by the application, but only by prevent_timer_throttling_tick()
	if (timer_handler==NULL) {
		return;
	}
    timer_handler();
}

void 
SystemJS_InstallTimerCallback(void (*timerHandler)())
{
    assert (timerHandler);
    timer_handler = timerHandler;
}

File: src/native/libs/System.Native.Browser/native/timer.ts
// Licensed to the .NET Foundation under one or more agreements.
// The .NET Foundation licenses this file to you under the MIT license.

import { } from "./cross-linked"; // ensure ambient symbols are declared

export function SystemJS_ScheduleTimer(shortestDueTimeMs: number): void {
    if (DOTNET.lastScheduledTimerId) {
        globalThis.clearTimeout(DOTNET.lastScheduledTimerId);
        Module.runtimeKeepalivePop();
        DOTNET.lastScheduledTimerId = undefined;
    }
    DOTNET.lastScheduledTimerId = safeSetTimeout(SystemJS_ScheduleTimerTick, shortestDueTimeMs);

    function SystemJS_ScheduleTimerTick(): void {
        DOTNET.lastScheduledTimerId = undefined;
        _SystemJS_ExecuteTimerCallback();
    }
}
SystemJS_ScheduleTimer["__deps"] = ["SystemJS_ExecuteTimerCallback"];

export function SystemJS_ScheduleBackgroundJob(): void {
    if (DOTNET.lastScheduledThreadPoolId) {
        globalThis.clearTimeout(DOTNET.lastScheduledThreadPoolId);
        Module.runtimeKeepalivePop();
        DOTNET.lastScheduledThreadPoolId = undefined;
    }
    DOTNET.lastScheduledThreadPoolId = safeSetTimeout(SystemJS_ScheduleBackgroundJobTick, 0);

    function SystemJS_ScheduleBackgroundJobTick(): void {
        DOTNET.lastScheduledThreadPoolId = undefined;
        _SystemJS_ExecuteBackgroundJobCallback();
    }
}
SystemJS_ScheduleBackgroundJob["__deps"] = ["SystemJS_ExecuteBackgroundJobCallback"];
